google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketPanel.java

```diff
@@ -0,0 +1,158 @@
+import com.intellij.openapi.project.Project;
+import com.intellij.ui.DocumentAdapter;
+import java.util.concurrent.Future;
+import javax.swing.DefaultListModel;
+import javax.swing.JLabel;
+import javax.swing.JList;
+import javax.swing.JPanel;
+import javax.swing.event.DocumentEvent;
+import org.apache.commons.lang.StringUtils;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * UI definition for the Google Cloud Storage panel. Displays the panel components and invokes the
+ * Storage API to load project buckets.
+ */
+final class GcsBucketPanel {
+
+  private JPanel gcsBucketPanel;
+  private ProjectSelector projectSelector;
+  private JPanel notificationPanel;
+  private JPanel bucketListPanel;
+  private JLabel notificationLabel;
+  private JList<Bucket> bucketList;
+  private DefaultListModel<Bucket> bucketListModel;
+  private Future<?> bucketLoadExecution;
+
+  GcsBucketPanel(@NotNull Project project) {
+    bucketListModel = new DefaultListModel<>();
+    bucketList.setModel(bucketListModel);
+    bucketList.setCellRenderer(new GcsBucketCellRenderer());
+    bucketList.setFixedCellHeight(25);
+    bucketList.setBackground(bucketListPanel.getBackground());
+
+    projectSelector
+        .getDocument()
+        .addDocumentListener(
+            new DocumentAdapter() {
+              @Override
+              protected void textChanged(DocumentEvent event) {
+                refresh();
+              }
+            });
+  }
+
+  void refresh() {
+    showNotificationPanel(true);
+
+    if (StringUtils.isEmpty(projectSelector.getText())) {
+      notificationLabel.setText(GctBundle.message("gcs.panel.bucket.listing.no.project.selected"));
+    } else {
+      String projectId = projectSelector.getText();
+      CredentialedUser user = projectSelector.getSelectedUser();
+
+      if (user != null) {
+        loadAndDisplayBuckets(projectId, user.getCredential());
+      } else {
+        notificationLabel.setText(
+            GctBundle.message("gcs.panel.bucket.listing.error.loading.buckets"));
+      }
+    }
+  }
+
+  @NotNull
+  JPanel getComponent() {
+    return gcsBucketPanel;
+  }
+
+  @VisibleForTesting
+  ProjectSelector getProjectSelector() {
+    return projectSelector;
+  }
+
+  @VisibleForTesting
+  JPanel getNotificationPanel() {
+    return notificationPanel;
+  }
+
+  @VisibleForTesting
+  JPanel getBucketListPanel() {
+    return bucketListPanel;
+  }
+
+  @VisibleForTesting
+  JLabel getNotificationLabel() {
+    return notificationLabel;
+  }
+
+  private void loadAndDisplayBuckets(String projectId, Credential credential) {
+    if (bucketLoadExecution != null && !bucketLoadExecution.isDone()) {
+      return;
+    }
+
+    bucketListModel.clear();
+    notificationLabel.setText(GctBundle.message("gcs.panel.bucket.listing.loading.text"));
+
+    bucketLoadExecution =
+        ApplicationManager.getApplication()
+            .executeOnPooledThread(
+                () -> {
+                  Storage storage =
```

Comment:
Is this supposed to not be retained from call-to-call? Or could this benefit from pulling it out into an instance variable?